src/ray/raylet/node_manager.cc

```diff
@@ -430,95 +429,58 @@ void NodeManager::ProcessClientMessage(
-      RAY_LOG(DEBUG) << "reconstructing object " << object_id;
       if (!task_dependency_manager_.CheckObjectLocal(object_id)) {
-        // TODO(swang): Instead of calling Pull on the object directly, record the
-        // fact that the blocked task is dependent on this object_id in the task
-        // dependency manager.
-        RAY_CHECK_OK(object_manager_.Pull(object_id));
+        if (message->fetch_only()) {
+          RAY_CHECK_OK(object_manager_.Pull(object_id));
+        } else {
+          // Add any missing objects to the list to subscribe to in the task
+          // dependency manager.
+          required_object_ids.push_back(object_id);
+        }
       }
+    }
 
-      if (!message->fetch_only()) {
-        // If the blocked client is a worker, and the worker isn't already blocked,
-        // then release any CPU resources that it acquired for its assigned task
-        // while it is blocked. The resources will be acquired again once the
-        // worker is unblocked.
-        std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
-        if (worker && !worker->IsBlocked()) {
-          RAY_CHECK(!worker->GetAssignedTaskId().is_nil());
-          auto tasks = local_queues_.RemoveTasks({worker->GetAssignedTaskId()});
-          const auto &task = tasks.front();
-          // Get the CPU resources required by the running task.
-          const auto required_resources =
-              task.GetTaskSpecification().GetRequiredResources();
-          double required_cpus = required_resources.GetNumCpus();
-          const std::unordered_map<std::string, double> cpu_resources = {
-              {kCPU_ResourceLabel, required_cpus}};
-
-          // Release the CPU resources.
-          auto const cpu_resource_ids = worker->ReleaseTaskCpuResources();
-          local_available_resources_.Release(cpu_resource_ids);
-          RAY_CHECK(cluster_resource_map_[gcs_client_->client_table().GetLocalClientId()]
-                        .Release(ResourceSet(cpu_resources)));
-
-          // Mark the task as blocked.
-          local_queues_.QueueBlockedTasks(tasks);
-          worker->MarkBlocked();
-
-          // Try to dispatch more tasks since the blocked worker released some
-          // resources.
-          DispatchTasks();
-        }
+    if (!required_object_ids.empty()) {
+      auto current_task_id = TaskID::nil();
+      std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
+      if (worker) {
+        RAY_CHECK(!worker->GetAssignedTaskId().is_nil());
+        current_task_id = worker->GetAssignedTaskId();
+
+        // Mark the worker as blocked.
+        HandleWorkerBlocked(worker);
       }
+      // Subscribe to the objects required by the ray.get. These objects will
+      // be fetched and/or reconstructed as necessary, until the objects become
+      // local or are unsubscribed.
+      task_dependency_manager_.SubscribeDependencies(current_task_id,
```

Comment:
What does it mean if this is called and `current_task_id` is nil?

Does that mean it's a driver? Won't that be an issue if there are multiple drivers?

Reply:
Yes, that means it's a driver.

Ah yes, I think you're right. We'll have to include the driver ID in the initial request to register a client with the raylet.